snowflake/exec.go

```diff
@@ -1,28 +1,73 @@
 package snowflake
 
 import (
+	"context"
 	"database/sql"
 
 	"github.com/jmoiron/sqlx"
+	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
-func ExecNoRows(db *sql.DB, query string) (sql.Result, error) {
-	logrus.Debug("[DEBUG] exec stmt ", query)
+func ExecNoRows(ctx context.Context, db *sql.DB, query string, args ...interface{}) (sql.Result, error) {
+	logrus.Debugf("[DEBUG] exec stmt (%s)", query)
 
-	return db.Exec(query)
+	// if no args, then don't prepare:
+	if len(args) == 0 {
+		return db.ExecContext(ctx, query)
+	}
+
+	// Prepare the statement
+	preparedStmt, err := db.PrepareContext(ctx, query)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Unable to prepare sql statement (%s)", query)
+	}
+	defer preparedStmt.Close()
+
+	sqlResult, err := preparedStmt.ExecContext(ctx, args...)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to execute prepared statement")
+	}
+
+	// Run statement with arguments
+	return sqlResult, nil
 }
 
-func ExecMulti(db *sql.DB, queries []string) error {
+func ExecMulti(ctx context.Context, db *sql.DB, queries []string, args ...[]interface{}) error {
 	logrus.Debug("[DEBUG] exec stmts ", queries)
 
-	tx, err := db.Begin()
+	tx, err := db.BeginTx(ctx, nil)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "Unable to begin transaction with database")
 	}
 
-	for _, query := range queries {
-		_, err = tx.Exec(query)
+	// NOTE(aku): do we need to store these statements one-by-one so we can programmatically close them all?
+	// preparedStmts := []*sql.Stmt{}
+
+	for i, query := range queries {
+		// Prepare the statement
+		logrus.Debugf("[DEBUG] exec (%s) ", query)
+		// if no args, then don't prepare:
+		if len(args[i]) == 0 {
+			_, err = db.ExecContext(ctx, query)
+			if err != nil {
+				return errors.Wrap(err, "Unable to execute statement")
+			}
+
+			return nil
```

Comment:
I think you would want this to be `continue` right? So it can continue the loop and execute more queries?
```suggestion
			continue
```